src/ir/child-typer.h

```diff
@@ -0,0 +1,1074 @@
+
+  void visitMemorySize(MemorySize* curr) {}
+
+  void visitMemoryGrow(MemoryGrow* curr) {
+    notePointer(&curr->delta, curr->memory);
+  }
+
+  void visitUnreachable(Unreachable* curr) {}
+
+  void visitPop(Pop* curr) {}
+
+  void visitRefNull(RefNull* curr) {}
+
+  void visitRefIsNull(RefIsNull* curr) { noteAnyReference(&curr->value); }
+
+  void visitRefFunc(RefFunc* curr) {}
+
+  void visitRefEq(RefEq* curr) {
+    Type eqref(HeapType::eq, Nullable);
+    note(&curr->left, eqref);
+    note(&curr->right, eqref);
+  }
+
+  void visitTableGet(TableGet* curr) { note(&curr->index, Type::i32); }
+
+  void visitTableSet(TableSet* curr) {
+    note(&curr->index, Type::i32);
+    note(&curr->value, wasm.getTable(curr->table)->type);
+  }
+
+  void visitTableSize(TableSize* curr) {}
+
+  void visitTableGrow(TableGrow* curr) {
+    note(&curr->value, wasm.getTable(curr->table)->type);
+    note(&curr->delta, Type::i32);
+  }
+
+  void visitTableFill(TableFill* curr) {
+    auto type = wasm.getTable(curr->table)->type;
+    note(&curr->dest, Type::i32);
+    note(&curr->value, type);
+    note(&curr->size, Type::i32);
+  }
+
+  void visitTableCopy(TableCopy* curr) {
+    note(&curr->dest, Type::i32);
+    note(&curr->source, Type::i32);
+    note(&curr->size, Type::i32);
+  }
+
+  void visitTry(Try* curr) {
+    for (auto& expr : curr->catchBodies) {
+      note(&expr, curr->type);
+    }
+  }
+
+  void visitTryTable(TryTable* curr) { note(&curr->body, curr->type); }
+
+  void visitThrow(Throw* curr) {
+    auto type = wasm.getTag(curr->tag)->sig.params;
+    assert(curr->operands.size() == type.size());
+    for (size_t i = 0; i < type.size(); ++i) {
+      note(&curr->operands[i], type[i]);
+    }
+  }
+
+  void visitRethrow(Rethrow* curr) {}
+
+  void visitThrowRef(ThrowRef* curr) {
+    note(&curr->exnref, Type(HeapType::exn, Nullable));
+  }
+
+  void visitTupleMake(TupleMake* curr) {
+    for (auto& expr : curr->operands) {
+      noteAny(&expr);
+    }
+  }
+
+  void visitTupleExtract(TupleExtract* curr,
+                         std::optional<size_t> arity = std::nullopt) {
+    if (!arity) {
+      assert(curr->tuple->type.isTuple());
+      arity = curr->tuple->type.size();
+    }
+    noteAnyTuple(&curr->tuple, *arity);
+  }
+
+  void visitRefI31(RefI31* curr) { note(&curr->value, Type::i32); }
+
+  void visitI31Get(I31Get* curr) {
+    note(&curr->i31, Type(HeapType::i31, Nullable));
+  }
+
+  void visitCallRef(CallRef* curr, std::optional<HeapType> ht = std::nullopt) {
+    if (!ht) {
+      ht = curr->target->type.getHeapType().getSignature();
+    }
+    auto params = ht->getSignature().params;
+    assert(curr->operands.size() == params.size());
+    for (size_t i = 0; i < params.size(); ++i) {
+      note(&curr->operands[i], params[i]);
+    }
+    note(&curr->target, Type(*ht, Nullable));
+  }
+
+  void visitRefTest(RefTest* curr) {
+    auto top = curr->castType.getHeapType().getTop();
+    note(&curr->ref, Type(top, Nullable));
```

Comment:
Similar question: This gets the top type and then verifies using that, but that will always succeed in validation.